lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -91,20 +91,24 @@ def capture(money, authorization, options = {})
         commit(action, post, options)
       end
 
-      def refund(money, authorization, options = {})
+      def refund(money, auth_or_payment_method, options = {})
         post = {}
         action = 'refund'
 
         add_datetime(post, options)
         add_invoice(post, money, options)
         add_clerk(post, options)
         add_transaction(post, options)
-        add_card(action, post, get_card_token(authorization), options)
+        auth_or_payment_method.is_a?(CreditCard) ? add_card(action, post, get_card_token(auth_or_payment_method), options) : add_card(action, post, auth_or_payment_method, options)
```

Comment:
💭 Opinion: Maybe instead of calling twice the add_card method, you can try something like this.

```ruby
card_token = auth_or_payment_method.is_a?(CreditCard) ? get_card_token(auth_or_payment_method) : auth_or_payment_method
add_card(action, post, card_token, options))


```

Reply:
thanks for the feedback, it looks cleaner that way. Change applied!